### PR TITLE
Merge the handlers to update engine and flutter status.

### PR DESCRIPTION
**Pull request by godofredoc**
This is in preparation to enable status updates for the flutter/plugin repository.

Bug: https://github.com/flutter/flutter/issues/84835

### CODE REVIEW

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/push_build_status_to_github.dart:55]: If you pass the full repo slug, you can do `Repo.fromFullName`

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/push_build_status_to_github.dart:66]:
```suggestion
```

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/push_build_status_to_github.dart:77]:
```suggestion
    for (String branch in config.getSupportedBranches(slug)) {
```

**Review by CaseyHillers — COMMENTED** [review]
Overall, this PR LGTM. There's a few blockers needing to be resolved before this can land (such as https://github.com/flutter/cocoon/pull/1204 and adding the engine to the supported scheduler repo to create the datastore entities)

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (godofredoc) [reply app_dart/lib/src/request_handlers/push_build_status_to_github.dart:55]: Not sure what Repo are you referring to I could find any reference to either Repo or fromFullName.

**Comment by godofredoc** [comment]
> Overall, this PR LGTM. There's a few blockers needing to be resolved before this can land (such as #1204 and adding the engine to the supported scheduler repo to create the datastore entities)
Good point, I didn't notice this logic was using the data models instead of the LUCI RPCs. I'll update it to use LUCI RPCs to remove the latency needed to inject the data and the current blocker.

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (godofredoc) [reply app_dart/lib/src/request_handlers/push_build_status_to_github.dart]: done

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (godofredoc) [reply app_dart/lib/src/request_handlers/push_build_status_to_github.dart]: done

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/push_build_status_to_github.dart:56]:
```suggestion
    scheduler.setLogger(log);
```

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/push_build_status_to_github.dart:69]: Should this be `GithubBuildStatusUpdate.statusFailure`?

If we have a 100 pending commits open, we should close the tree.

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (CaseyHillers) [inline app_dart/lib/src/request_handlers/push_build_status_to_github.dart:71]: Can this be implemented in the BuildStatusProvider https://cs.opensource.google/flutter/cocoon/+/master:app_dart/lib/src/service/build_status_provider.dart (LUCI RPCs vs Datastore)

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (godofredoc) [reply app_dart/lib/src/request_handlers/push_build_status_to_github.dart:69]: It is already limited to 40 states per builder and handled in _getLatestStatus

File: app_dart/test/request_handlers/push_build_status_to_github_test.dart
Comment (keyonghan) [inline app_dart/test/request_handlers/push_build_status_to_github_test.dart:26]: shall we add a test to update status for multiple branches in flutter/flutter?

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/push_build_status_to_github.dart:88]: This will list PRs of master branch only, and will repeat same updates within the branch loop.

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/push_build_status_to_github.dart:116]: Same here. This omits other branches data.

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (godofredoc) [reply app_dart/lib/src/request_handlers/push_build_status_to_github.dart:71]: Moved to the service class.

File: app_dart/test/request_handlers/push_build_status_to_github_test.dart
Comment (godofredoc) [reply app_dart/test/request_handlers/push_build_status_to_github_test.dart:26]: Actually the reason I didn't add tests for different branches is that the more I look at the code the more convinced I am that we don't need tree status for branches.

@christopherfujino do we actually need branch support for tree status?

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (godofredoc) [reply app_dart/lib/src/request_handlers/push_build_status_to_github.dart:116]: Looking at the code we may actually want to remove the complexity associated to branches I don't see any current use cases for it.

Comment (godofredoc) [reply app_dart/lib/src/request_handlers/push_build_status_to_github.dart:116]: I'll wait for @christopherfujino reply, if there are release use cases for tree status in branches I'll add relevant tests.

File: app_dart/test/request_handlers/push_build_status_to_github_test.dart
Comment (christopherfujino) [reply app_dart/test/request_handlers/push_build_status_to_github_test.dart:26]: we don't need this.

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (christopherfujino) [reply app_dart/lib/src/request_handlers/push_build_status_to_github.dart]: yeah, deleting this is fine

File: app_dart/lib/src/service/build_status_provider.dart
Comment (CaseyHillers) [inline app_dart/lib/src/service/build_status_provider.dart:132]: This looks at all LUCI builds in the number of commits * number of builders. We shouldn't be checking builds if they have a green build above them. Should we refactor calculateCumulativeStatus so it injects `tasks` and we can reuse that algorithm?

File: app_dart/lib/src/service/build_status_provider.dart
Comment (CaseyHillers) [inline app_dart/lib/src/service/build_status_provider.dart:149]: I'm concerned defaulting to this will make it harder to rename builders

File: app_dart/lib/src/service/build_status_provider.dart
Comment (godofredoc) [reply app_dart/lib/src/service/build_status_provider.dart:132]: That will add back the latency to ingest the data. I would prefer to optimize this later with its own data model.

File: app_dart/lib/src/service/build_status_provider.dart
Comment (godofredoc) [reply app_dart/lib/src/service/build_status_provider.dart:149]: Can you explain why?

Comment (godofredoc) [reply app_dart/lib/src/service/build_status_provider.dart:149]: Please note that this will be taking into account only active builders as returned by scheduler.

Comment (CaseyHillers) [reply app_dart/lib/src/service/build_status_provider.dart:149]: Looking at it again, it shouldn't be a concern. For example, in flutter/flutterI need to rename `framework_tests_misc` to `Windows framework_tests_misc` to make the name match the builder name. However, this is aggregating by the builder name for now.

File: app_dart/test/request_handlers/push_build_status_to_github_test.dart
Comment (godofredoc) [reply app_dart/test/request_handlers/push_build_status_to_github_test.dart:26]: thanks, removed.

**Review by CaseyHillers — APPROVED** [review]
LGTM. Feel free to run a prod traffic test to validate it e2e.

Do we need to land this when both trees are green? Or will it be able to update the old status checks?

File: app_dart/lib/src/request_handlers/push_build_status_to_github.dart
Comment (keyonghan) [reply app_dart/lib/src/request_handlers/push_build_status_to_github.dart]: This much simplifies the logic!

**Comment by godofredoc** [comment]
> LGTM. Feel free to run a prod traffic test to validate it e2e.
> 
> Do we need to land this when both trees are green? Or will it be able to update the old status checks?

It will also update old status checks. How do we land changes on the cron jobs config? is the deployment script updating the crons or that part is still a manual process.

I'll validate the change outside of working hours to not affect anyone and land it if successful.

**Review by keyonghan — APPROVED** [review]
LGTM.

**Comment by CaseyHillers** [comment]
> It will also update old status checks. How do we land changes on the cron jobs config? is the deployment script updating the crons or that part is still a manual process.

See https://github.com/flutter/cocoon/tree/master/app_dart#updating-cloud-cron-tasks

**Pull request merged**
